proto/document/v1/document.proto

```diff
@@ -0,0 +1,126 @@
+syntax = "proto3";
+package nitric.document.v1;
+
+import "google/protobuf/struct.proto";
+
+// protoc plugin options for code generation
+option go_package = "nitric/v1;v1";
+option java_package = "io.nitric.proto.document.v1";
+option java_multiple_files = true;
+option java_outer_classname = "Documents";
+option php_namespace = "Nitric\\Proto\\Document\\V1";
+option csharp_namespace = "Nitric.Proto.Document.v1";
+
+// Service for storage and retrieval of simple JSON keyValue
+service Document {
+  // Get an existing document
+  rpc Get (DocumentGetRequest) returns (DocumentGetResponse);
+
+  // Create a new or overwrite an existing document
+  rpc Set (DocumentSetRequest) returns (DocumentSetResponse);
+
+  // Delete an existing document
+  rpc Delete (DocumentDeleteRequest) returns (DocumentDeleteResponse);
+  
+  // Query the document collection (pagination support)
+  rpc Query (DocumentQueryRequest) returns (DocumentQueryResponse);
+  
+  // Query All the document collection (streaming support)
+  rpc QueryAll (DocumentQueryAllRequest) returns (stream DocumentQueryAllResponse);
+}
+
+message Key {
+  // The item collection
+  string collection = 1;
+
+  // The items unique id
+  string id = 2;
+}
+
+message DocumentGetRequest {
+  // The document root key
+  Key key = 1;
+
+  // The optional sub collection key
+  Key sub_key = 2;
+}
+
+message DocumentGetResponse {
+  // The retrieved value
+  google.protobuf.Struct value = 1;
+}
+
+message DocumentSetRequest {
+  // The document root key
+  Key key = 1;
+
+  // The optional sub collection key
+  Key sub_key = 2;
+
+  // A simple JSON object to store
+  google.protobuf.Struct value = 3;
+}
+
+message DocumentSetResponse {}
+
+message DocumentDeleteRequest {
+  // The document root key
+  Key key = 1;
+
+  // The optional sub collection key
+  Key sub_key = 2;
+}
+
+message DocumentDeleteResponse {}
+
+message Expression {
+  // The query operand or attribute
+  string operand = 1;
+
+  // The query operator [ = | < | <= | > | >= | startsWith ]
+  string operator = 2;
+
+  // The query expression value
+  string value = 3;
+}
+
+message DocumentQueryRequest {
+  // The query root key.collection, and optional parent key.id for subcollection queries
+  Key key = 1;
```

Comment:
Should the parameters here be collection (not optional) and an optional partition/parent key?

```
string collection = 1;
// optional parent key, used for scoping sub-collection queries
Key parent_key = 2;
```

Reply:
In the implementation of this feature, I have found using switching the contexts around to be quite confusing in the code, e.g. the context is a top level collection and the parent is undefined, or the context is a subcollection and the parent is defined. In these different contexts the key could either be a top level key or a sub collection key. This pattern makes the Query API more similar to the other Get/Set/Delete APIs